enterprise/frontend/src/embedding-sdk/store/auth.ts

```diff
@@ -0,0 +1,158 @@
+
+    if (isValidJwtConfig) {
+      // JWT setup
+      api.onBeforeRequest = async () => {
+        const session = await dispatch(
+          getOrRefreshSession(sdkConfig.jwtProviderUri!),
+        ).unwrap();
+        if (session?.id) {
+          api.sessionToken = session.id;
+        }
+      };
+      // verify that the session is actually valid before proceeding
+      await dispatch(getOrRefreshSession(sdkConfig.jwtProviderUri!)).unwrap();
+    } else if (isValidApiKeyConfig) {
+      // API key setup
+      api.apiKey = sdkConfig.apiKey;
+    }
+    // Fetch user and site settings
+    const [user, siteSettings] = await Promise.all([
+      dispatch(refreshCurrentUser()),
+      dispatch(refreshSiteSettings({})),
+    ]);
+
+    if (!user.payload) {
+      // The refresh user thunk just returns null if it fails to fetch the user, it doesn't throw
+      const error = new Error(
+        "Failed to fetch the user, is the session valid?",
+      );
+      console.error(error);
+      throw error;
+    }
+    if (!siteSettings.payload) {
+      const error = new Error(
+        "Failed to fetch the site settings, is the session valid?",
+      );
+      console.error(error);
+      throw error;
+    }
+  },
+);
+
+export const refreshTokenAsync = createAsyncThunk(
+  "sdk/token/REFRESH_TOKEN",
+  async (url: string, { getState }): Promise<EmbeddingSessionToken | null> => {
+    // The SDK user can provide a custom function to refresh the token.
+    const customGetRefreshToken = getFetchRefreshTokenFn(
+      getState() as SdkStoreState,
+    );
+
+    const getRefreshToken = customGetRefreshToken ?? defaultGetRefreshTokenFn;
+
+    // # How does the error handling work?
+    // This is an async thunk, thunks _by design_ can fail and no error will be shown on the console (it's the reducer that should handle the reject action)
+    // The following lines are wrapped in a try/catch block that will catch any errors thrown, log them to the console as a big red errors, and re-throw them to make the thunk reject
+    // In this way we also support standard thrown Errors in the custom fetchRequestToken user provided function
+
+    try {
+      const response = await getRefreshToken(url);
+      const source = customGetRefreshToken
+        ? '"fetchRequestToken"'
+        : "jwtProvider endpoint";
+
+      if (!response || typeof response !== "object") {
+        throw new Error(
+          `The ${source} must return an object with the shape {id:string, exp:number, iat:number, status:string}, got ${safeStringify(response)} instead`,
+        );
+      }
+      if ("status" in response && response.status !== "ok") {
+        if ("message" in response && typeof response.message === "string") {
+          // For some errors, the BE gives us a message that explains it
+          throw new Error(response.message);
+        }
+        if ("status" in response && typeof response.status === "string") {
```

Comment:
the first condition isn't needed, because this one is already nested in a condition that contains it.

```suggestion
        if (typeof response.status === "string") {
```

Reply:
I hate this name, `response.status` usually means [HTTP response](https://developer.mozilla.org/en-US/docs/Web/API/Response/status) status as a number.

In this case, this seems to be something we thrown from BE.
![Screenshot 2024-10-30 at 10 37 25 PM](https://github.com/user-attachments/assets/7c0a27eb-935b-4022-a4c1-fbf1024e002b)

Reply:
that would work. I don't know if we have a type for this but if we do that will also help, this feels like there should be a type for it. Like MetabaseSession or something